### ISSUE TITLE
Should OP-TEE revision be printed as "string" in linux OP-TEE driver ?

### ISSUE THREAD

**Issue by sahilnxp** [OP-TEE/optee_os#7141]
Since we have moved to [16bit sha1 len](https://github.com/OP-TEE/optee_os/blob/master/core/core.mk#L16) for OP-TEE revision in case of 64 bit architecture.

Should we print the revision as `string` in place of  `long int`  [here](https://github.com/torvalds/linux/blob/master/drivers/tee/optee/smc_abi.c#L1275)

Observed one issue on 64 bit architectures, if the commit id is starting with `0` like `04d1c612ec7beaede073b8cad0f33a1f5ab9e2bc`

It is printing revision as below removing leading 0.
`2024-11-15T02:29:51.469317 [    2.019585] optee: revision 4.4 (4d1c612ec7beaed)`

**Comment by jforissier** [comment]
Hi @sahilnxp,

> Should we print the revision as `string` in place of `long int` [here](https://github.com/torvalds/linux/blob/master/drivers/tee/optee/smc_abi.c#L1275)

I think the format simply needs adjusting to cater to the size of the unsigned long:

```C
pr_info("revision %lu.%lu (%0*lx)", res.result.major,
	res.result.minor, sizeof(res.result.build_id) * 2,
	res.result.build_id);
```

**Comment by sahilnxp** [comment]
> Hi @sahilnxp,
> 
> > Should we print the revision as `string` in place of `long int` [here](https://github.com/torvalds/linux/blob/master/drivers/tee/optee/smc_abi.c#L1275)
> 
> I think the format simply needs adjusting to cater to the size of the unsigned long:
> 
> ```c
> pr_info("revision %lu.%lu (%0*lx)", res.result.major,
> 	res.result.minor, sizeof(res.result.build_id) * 2,
> 	res.result.build_id);
> ```

Yes, that will work, but we need to pass the size like (16) too in this case like below:

```
pr_info("revision %lu.%lu (%0*lx)", res.result.major,
	res.result.minor, sizeof(res.result.build_id) * 2,
	16, res.result.build_id);
```
and how to decide on this size ? Otherwise, it is giving the following warning.

```
                 from drivers/tee/optee/smc_abi.c:12:
drivers/tee/optee/smc_abi.c: In function ‘optee_msg_get_os_revision’:
./include/linux/kern_levels.h:5:25: warning: field width specifier ‘*’ expects argument of type ‘int’, but argument 4 has type ‘long unsigned int’ [-Wformat=]
    5 | #define KERN_SOH        "\001"          /* ASCII Start Of Header */
      |                         ^~~~~~

```

**Comment by jenswi-linaro** [comment]
Casting the size argument to `int` should work.

**Comment by sahilnxp** [comment]
Sorry for confusion here, 

There is no compilation warning in case of following code:

```
pr_info("revision %lu.%lu (%0*lx)", res.result.major,
	res.result.minor, sizeof(res.result.build_id) * 2,
	16, res.result.build_id);
```

But how we can decide if we should give the size as 8 (for 32bit arch) and 16 (for 64 bit arch) ?

**Comment by jforissier** [comment]
> Sorry for confusion here.
> 
> There is no compilation warning in case of following code:
> 
> ```
> pr_info("revision %lu.%lu (%0*lx)", res.result.major,
> 	res.result.minor, sizeof(res.result.build_id) * 2,
> 	16, res.result.build_id);
> ```
> 
> But how we can decide if we should give the size as 8 (for 32bit arch) and 16 (for 64 bit arch) ?

`res.result.build_id` is `unsigned long`. 32-bit arch have `sizeof(unsigned long) == 4` while 64-bit arch have `sizeof(unsigned long) == 8`.

**Comment by sahilnxp** [comment]
So it should be like this, right ?

```
pr_info("revision %lu.%lu (%0*lx)", res.result.major,
             res.result.minor, (int)sizeof(unsigned long),
             res.result.build_id);
```

**Comment by jforissier** [comment]
No, because in the Linux (and OP-TEE) coding style, the argument of `sizeof()` should be the variable rather than the type. This makes sense to make maintenance easier in case the type is changed for instance.

**Comment by sahilnxp** [comment]
In that case, it should be like below ?

```
pr_info("revision %lu.%lu (%0*lx)", res.result.major,
             res.result.minor, (int)sizeof(res.result.build_id),
             res.result.build_id);
```

**Comment by jforissier** [comment]
> In that case, it should be like below ?
> 
> ```
> pr_info("revision %lu.%lu (%0*lx)", res.result.major,
>              res.result.minor, (int)sizeof(res.result.build_id),
>              res.result.build_id);
> ```

No. The `* 2` is missing. Here is exactly how the code should be:

```C
	if (res.result.build_id)
		pr_info("revision %lu.%lu (%08lx)", res.result.major,
			res.result.minor, (int)sizeof(res.result.build_id) * 2,
			res.result.build_id);
	else
		pr_info("revision %lu.%lu", res.result.major, res.result.minor);
 ```

**Comment by jforissier** [comment]
Because for each byte of data, two characters are printed in hex

**Comment by sahilnxp** [comment]
> Because for each byte of data, two characters are printed in hex

ohh, yeah. Thats right. I missed that part. Thanks.
I will send a patch for this on Linux kernel and update link here.

**Comment by sahilnxp** [comment]
Patch has been sent upstream: https://lore.kernel.org/linux-kernel/20241129114648.3048941-1-sahil.malhotra@nxp.com/

**Comment by sahilnxp** [comment]
>  https://lore.kernel.org/linux-kernel/20241129114648.3048941-1-sahil.malhotra@nxp.com/

@jforissier Do I need to update the tags in this one too ?

**Comment by jforissier** [comment]
> > https://lore.kernel.org/linux-kernel/20241129114648.3048941-1-sahil.malhotra@nxp.com/
> 
> @jforissier Do I need to update the tags in this one too ?

I don't think one needs to re-send patches to the LKML just for the tags because the maintainers have different means to collect them. @jenswi-linaro is this correct?

**Comment by jenswi-linaro** [comment]
That's right, I'll apply all tags (and add my Signed-off-by) when I pick it up. It's a few weeks until the next merge window opens so there's no rush.

**Comment by sahilnxp** [comment]
Thanks @jenswi-linaro @jforissier for your inputs.
